libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TemplatesRegisterExtensions.cs

```diff
@@ -0,0 +1,195 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using AdaptiveExpressions;
+using AdaptiveExpressions.Memory;
+using Microsoft.Bot.Builder.Dialogs.Adaptive.Generators;
+using Microsoft.Bot.Builder.LanguageGeneration;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive
+{
+    /// <summary>
+    /// Register lg templates into Expression.
+    /// Using prefix lg. to avoid the conflict with pre-built functions.
+    /// </summary>
+    public static class TemplatesRegisterExtensions
+    {
+        /// <summary>
+        /// Register Template functions.
+        /// </summary>
+        /// <param name="dialogContext">Dialog context.</param>
+        /// <param name="generator">Language generator.</param>
+        public static void RegisterTemplateFunctions(this DialogContext dialogContext, LanguageGenerator generator)
+        {
+            var lgm = dialogContext.Services.Get<LanguageGeneratorManager>();
+            var templateNames = GetTemplates(generator, dialogContext).Select(u => u.Name).Distinct();
+
+            var prebuildFunctionNames = Expression.Functions.Values.Select(u => u.Type);
+            foreach (var templateName in templateNames)
+            {
+                var functionType = prebuildFunctionNames.Contains(templateName) ? "lg." + templateName : templateName;
```

Comment:
>"lg." [](start = 82, length = 5)

LG today requires all template names in scope to be globally unique so this serves to distinguish template names from prebuilts.  What happens when template names can use namespaces?  If I do import foo = Microsoft.Bot.Community#woof.en-us.lg then inside the file you would refer to foo.Welcome().  What would be the name here?  Would it be lg.foo.Welcome() and if imported without namespace lg.Welcome()?

Reply:
So, i'm looking at the internal lg naming conflicting issue as a decoupled with the lg names have conflicts with existing dialog functions. 

The namespacing proposal like "foo.welcome" is a notion to handle the former case. With that been applied, the internal LG names have become unique with a namespace prefix, like foo.welcome. 

This "lg" prefix is a notion to handle the latter case. Where, if "foo.welcome" still conflicts with built-in functions or customized functions(which is rarely the case), you need to refer it to "lg.foo.welcome".

Reply:
Correction on my previous reply. After taking a deeper look at this with @Danieladu , we think it makes more sense to only expose the direct templates from the the lg file referred by the dialog, without dealing with this lg's files imports (regardless it's imported directly or via an alais). 

That is more clean and more modular, and works in the same mechanism as programming languages, ie, when a dialog import an lg, it brings in all it's templates, but NOT it's lg's imports.  And this will not lose any capability since, you can also wrap the imported into a direct template.

Reply:
Currently, the templates in default import(Personal think static import is more accurate) would be injected into expression, but the alias import would not. For example:
a.lg
```
[](b.lg)
# t1
- in a, call b ${t2()}
```

b.lg
```
# t2
- in b
```

`t1` and `t2` would both be injected into expression.

But for the alias import:
a.lg
```
[](b.lg) as base
# t1
- in a, call b ${base.t2()}
```

b.lg
```
# t2
- in b
```

`t1` would be injected into expression, but `t2` or `base.t2` would not.
Notice: In declarative dialog, all LG files would be loaded and parsed, so, finally, `t2` would be injected into expression too.

Reply:
I don't think that works.  The dialog file only has a single LG file--how would I refer to templates that come from 2 different files?  I would be forced to wrap every one?  This doesn't seem like a good solution.  

---
In reply to: [614484772](https://github.com/microsoft/botbuilder-dotnet/pull/5319#discussion_r614484772) [](ancestors = 614484772)

Reply:
In this case, 
a.lg
```
[](b.lg) as base
# t1
- in a, call b ${base.t2()}
```


b.lg
```
# t2
- in b
```
You mean we should also export `base.t2` that's right? It is make sense. But `t2` should not I think.

Reply:
I have changed the policy to inject both `t1` and `base.t2`, please take a look.

Reply:
Sounds good to me. Thanks!